src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -77,55 +77,61 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+        if (userID == null || recipientID == null) {
             response.sendError(500);
         } else {
             Query query = new Query("user");
             query.setFilter(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, userID));
 
             Entity userEntity = database.prepare(query).asSingleEntity();
-            
             User user = new User(userEntity);
-            PreparedQuery results = DatastoreServiceFactory.getDatastoreService().prepare(query);
-
 
             Gson gson = new Gson();
-
             response.setContentType("application/json");
             response.getWriter().println(gson.toJson(user));
         }
     }
 
     @Override
     public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        String jsonString = IOUtils.toString((request.getInputStream()));
-        User userInput = new Gson().fromJson(jsonString, User.class);
 
-        String userID = userInput.userId;
-        
-        // Entity with a userID as identifier.
-        Entity newUser = new Entity("user", userID);
-        newUser.setProperty("userId", userInput.userId);
-        newUser.setProperty("name", userInput.name);
-        newUser.setProperty("email", userInput.email);
-        newUser.setProperty("language", userInput.language);
-        database.put(newUser);
+        String userID = request.getHeader("userId");
+
+        if (userID == null || recipientID == null) {
+            response.sendError(500);
+        } else {
+            String jsonString = IOUtils.toString((request.getInputStream()));
+            User userInput = new Gson().fromJson(jsonString, User.class);
+
+            // Entity with a userID as identifier.
+            Entity newUser = new Entity("user", newPost.userId);
+            newUser.setProperty("userId", newPost.userId);
+            newUser.setProperty("name", userInput.name);
+            newUser.setProperty("email", userInput.email);
+            newUser.setProperty("language", userInput.language);
+            database.put(newUser);
+        }
     }
 
     @Override
     public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
-        String jsonString = IOUtils.toString((request.getInputStream()));
-        User userInput = new Gson().fromJson(jsonString, User.class);
-        String userID = userInput.userId;
-
-        // Update user using userId as entity identifier.
-        Entity userToUpdate = new Entity("user", userID);
-        userToUpdate.setProperty("userId", userInput.userId);
-        userToUpdate.setProperty("name", userInput.name);
-        userToUpdate.setProperty("email", userInput.email);
-        userToUpdate.setProperty("language", userInput.language);
-
-        database.put(userToUpdate);
-    }
 
+        String userID = request.getHeader("userId");
+
+        if (userID == null || recipientID == null) {
+            response.sendError(500);
```

Comment:
i think this response should be handled in the AuthFilter.java servlet, because if the userID is null, it means the idToken attacked to the request header and the call was not able to be decoded successfully in AuthFilter.java

I could work to make that happen in AuthFilter.java if you think that is a better approach ??

Reply:
This should also be kind of outdated. I only check if the userID is null. While it would be rejected in the AuthFilter, it's just an extra step to secure the app. 

We could easily deprecate it however

Reply:
oh ok. i think it is fine to check if the userId is null